autorest/azure/devicetoken.go

```diff
@@ -0,0 +1,196 @@
+}
+
+// TokenError is the object returned by the token exchange endpoint
+// when something is amiss
+type TokenError struct {
+	Error            *string `json:"error"`
+	ErrorCodes       []int   `json:"error_codes"`
+	ErrorDescription *string `json:"error_description"`
+	Timestamp        *string `json:"timestamp"`
+	TraceID          *string `json:"trace_id"`
+}
+
+// DeviceToken is the object return by the token exchange endpoint
+// It can either look like a Token or an ErrorToken, so put both here
+// and check for presence of "Error" to know if we are in error state
+type DeviceToken struct {
+	Token
+	TokenError
+}
+
+// InitiateDeviceAuth initiates a device auth flow. It returns a DeviceCode
+// that can be used with CheckForUserCompletion or WaitForUserCompletion.
+func InitiateDeviceAuth(client *autorest.Client, clientID, resource string) (*DeviceCode, error) {
+	req, _ := autorest.Prepare(
+		&http.Request{},
+		autorest.AsPost(),
+		autorest.WithBaseURL(OAuthDeviceEndpoint),
+		autorest.WithFormData(url.Values{
+			"client_id": []string{clientID},
+			"resource":  []string{resource},
+		}),
+		autorest.WithQueryParameters(map[string]interface{}{
+			authAPIVersionQueryParamName: authAPIVersionQueryParamValue,
+		}))
+
+	resp, err := client.Send(req)
+	if err != nil {
+		return nil, fmt.Errorf("Error occurred while requesting Device Authorization Code. %s", err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf(
+			"Request for Device Authorization Code resulted in Status: %d: %s",
+			resp.StatusCode,
+			http.StatusText(resp.StatusCode))
+	}
+
+	var code DeviceCode
+	decoder := json.NewDecoder(resp.Body)
+	err = decoder.Decode(&code)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to decode Device Authorization Code json")
+	}
+
+	code.Resource = &resource
+
+	return &code, nil
+}
+
+// CheckForUserCompletion takes a DeviceCode and checks with the Azure AD OAuth endpoint
+// to see if the device flow has: been completed, timed out, or otherwise failed
+func CheckForUserCompletion(client *autorest.Client, clientID string, code *DeviceCode) (*Token, error) {
+	req, _ := autorest.Prepare(
+		&http.Request{},
+		autorest.AsPost(),
+		autorest.WithBaseURL(OAuthTokenEndpoint),
+		autorest.WithFormData(url.Values{
+			"client_id":  []string{clientID},
+			"code":       []string{*code.DeviceCode},
+			"grant_type": []string{OAuthGrantTypeDeviceCode},
+			"resource":   []string{*code.Resource},
+		}),
+		autorest.WithQueryParameters(map[string]interface{}{
+			authAPIVersionQueryParamName: authAPIVersionQueryParamValue,
+		}))
+
+	resp, err := client.Send(req)
+	if err != nil {
+		return nil, fmt.Errorf("Error occurred while exchanging device code for token. %s", err)
+	}
+	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusBadRequest {
+		return nil, fmt.Errorf(
+			"Exchanging Device Code for OAuth token resulted in Status: %d: %s",
+			resp.StatusCode,
+			http.StatusText(resp.StatusCode))
+	}
```

Comment:
yeah this status code check + body close + decoding is repeated in two methods, we can use response decorators here.